lib/kernel/src/gen_tcp.erl

```diff
@@ -161,7 +161,7 @@ connect(SockAddr, Opts) ->
     connect(SockAddr, Opts, infinity).
 
 -spec connect(Address, Port, Opts) -> {ok, Socket} | {error, Reason} when
-      Address  :: inet:socket_address() | inet:hostname(),
+      Address  :: inet:socket_address() | nonempty_string() | atom(),
```

Comment:
Should we not change `inet:hostname() `to -`type hostname() :: atom() | nonempty_string().` rather that changing this spec? @RaimoNiskanen ?

Reply:
We should, if that does not break anything :)

Reply:
Since changing the `inet:hostname()` may involve further changes to handle other cases, do we accept this simple change or not?

Reply:
AFAIK, empty string was not allowed but an empty atom is. 
There were some edge cases here...

Reply:
I did some studying, acquired a new opinion, and wrote it in issue #6353, which was shortly that empty atom should work just like empty string, and since we cannot use the type language to exclude both we should allow both, hence we should not do this type change.

Furthermore I think that we should consistently change so that the empty host name is treated just like the host name "." i.e the root domain.  This would be consistent in the way that all other (non-empty) host names are always treated as absolute and have an implied dot at the end (in DNS's terminology - it is the way to denote an absolute host name in a DNS domain file and when quering to a DNS resolver).

For example `gen_tcp:connect/*` looks up an address for a host name, that is, an A or AAAA record.  And the root domain today has no address records (however e.g a SOA record).  So such a lookup should result in `{error, nxdomain}`, now, and probably in the future.  The right thing (tm) would be to do the lookup...